firmware/src/networking/protocol/serialize.rs

```diff
@@ -0,0 +1,27 @@
+#![allow(dead_code)]
+
+/// A simpler alternative to [`SerializeExact`], which will serialize without needing an
+/// exact buffer size.
+pub trait Serialize {
+	type Error;
+	/// Serializes into `buf`, returning the number of bytes written, or an error.
+	/// Note that this must not return `Ok` if only part of `self` was actually serialized.
+	fn serialize(&mut self, buf: &mut [u8]) -> Result<usize, Self::Error>;
+}
+
+/// Serializes directly into a buffer, such as directly into the tcp/udp buffers with no
+/// the exact size known.
+pub trait SerializeExact {
+	type Error;
+	/// Serializes the packet into the provided buffer. Implementations may choose to steal
+	/// `self`, or perform a copy, or they may have already serialized into bytes preemptively
+	/// (such as with flatbuffers).
+	///
+	/// # Panics
+	/// May panic if `f` returns an Ok variant with a buffer that is not the exact size as
+	/// `f`'s argument.
+	fn serialize_direct<'a, 'b>(
```

Comment:
We do not want to allocate at runtime, and UDP buffer size will not change over the course of runtime. I don't see the need for this interface. Maybe a `const Packet::MAX_SERIALIZED_LEN: usize` could be useful for reserving buffer space.

Reply:
the reason I used the term "reserve" was to avoid making it seem like this was allocating. Essentially, this trait's api arose from this function which I found:
https://docs.rs/smoltcp/0.8.2/smoltcp/socket/struct.UdpSocket.html#method.send

The important thing to note here, is that this function lets us serialize *directly* into the message payload, instead of to an intermediate buffer that gets copied in later. But it needs the exact # of bytes. And reserving that buffer may also fail.

So we needed an api that:
* Is tolerant to reservation failure, returning to the caller when that happens
* Can serialize directly into reserved buffer if the size of the final payload can be computed before the serialization is actually performed
* Does not panic under correct implementation.

This api does that. I'll give two examples in discord so we can discuss more there.